include/onnxruntime/core/framework/tensor.h

```diff
@@ -63,14 +77,37 @@ class Tensor final {
   /**
    * Create tensor with given type, shape, pre-allocated memory and allocator info.
    * This function won't check if the preallocated buffer(p_data) has enough room for the shape.
-   * \param data A preallocated buffer. Can be NULL if the shape is empty.
+   * \param p_data A preallocated buffer. Can be NULL if the shape is empty.
    *              Tensor does not own the data and will not delete it
    * \param alloc Where the buffer('data') was allocated from
    * \param offset Offset in bytes to start of Tensor within p_data. 
    */
   Tensor(MLDataType p_type, const TensorShape& shape, void* p_data, const OrtMemoryInfo& alloc,
          ptrdiff_t offset = 0);
 
+  /**
+   * Create tensor with given type, shape, pre-allocated memory and ways to release buffer.
+   * 
+   * This function won't check if the preallocated buffer(p_data) has enough room for the shape.
+   *
+   * \param p_data   A preallocated buffer. Can be NULL if the shape is empty.
+   *                 Must be allocated by the deleter, will be released when tensor destructs.
+   * \param deleter  points to the allocator from which p_data is allocated.
+   */
+  Tensor(MLDataType p_type, const TensorShape& shape, void* p_data, AllocatorPtr&& deleter);
+
+  /**
+   * Create tensor with given type, shape, pre-allocated memory and the allocator. 
+   * This version transfer the buffer ownership to the tensor.
+   * This function won't check if the preallocated buffer(p_data) has enough room for the shape.
+   * \param data A preallocated buffer. Can be NULL if the shape is empty.
+   *              Tensor ownes the data and will delete it
+   * \param offset Offset in bytes to start of Tensor within p_data. 
+   * \param deleters  functions for cleaning up and releasing buffer
+  */
+  Tensor(MLDataType p_type, const TensorShape& shape, void* p_data, ptrdiff_t offset, 
+         const OrtMemoryInfo& alloc, std::vector<std::function<void(void)>>&& deleters);
+
```

Comment:
This feels overly complicated just to add ownership given we already have a ctor that takes an allocator. Could we instead create a Tensor in the ITensorAllocator instead of a MemBuffer, as I'm not sure what value MemBuffer is adding.

Reply:
that is not enough when we have two different buffers to free. As of why a Tensor holds two different buffers, it does seems to me a big memory waste, but I need to figure out the reason behind it before changing it.